snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import uuid
+from typing import Callable, Optional, TypeVar
+
+from pkg_resources import resource_string
+from redis.exceptions import ResponseError
+
+from snuba.redis import RedisClientType
+from snuba.state import get_config
+from snuba.state.cache.abstract import Cache, ExecutionError, ExecutionTimeoutError
+from snuba.utils.codecs import Codec
+
+
+T = TypeVar("T")
+
+
+logger = logging.getLogger(__name__)
+
+
+RESULT_VALUE = 0
+RESULT_EXECUTE = 1
+RESULT_WAIT = 2
+
+
+class RedisCache(Cache[T]):
+    def __init__(
+        self, client: RedisClientType, prefix: str, codec: Codec[bytes, T]
+    ) -> None:
+        self.__client = client
+        self.__prefix = prefix
+        self.__codec = codec
+
+        # TODO: This should probably be lazily instantiated, rather than
+        # automatically happening at startup.
+        self.__script_get = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/get.lua")
+        )
+        self.__script_set = client.register_script(
+            resource_string("snuba", "state/cache/redis/scripts/set.lua")
+        )
+
+    def __build_key(self, key: str) -> str:
+        return f"{self.__prefix}{{{key}}}"
+
+    def get(self, key: str) -> Optional[T]:
+        value = self.__client.get(self.__build_key(key))
+        if value is None:
+            return None
+
+        return self.__codec.decode(value)
+
+    def set(self, key: str, value: T) -> None:
+        self.__client.set(
+            self.__build_key(key),
+            self.__codec.encode(value),
+            ex=get_config("cache_expiry_sec", 1),
+        )
+
+    def get_or_execute(self, key: str, function: Callable[[], T], timeout: int) -> T:
+        result_key = self.__build_key(key)
+        wait_queue_key = f"{self.__build_key(key)}:w"
+        task_ident_key = f"{self.__build_key(key)}:t"
+
+        result = self.__script_get(
+            [result_key, wait_queue_key, task_ident_key], [timeout, uuid.uuid1().hex]
```

Comment:
I may be missing something but I don't get why we need a unique task identity.
This task identity is set as value of task_ident_key in cache when the queue for a specific result_key is created. Everyone who gets enqueued after the first get_or_execute reuses the same returned by the get script, which is the one stored when the queue was created.
Isn't the cardinality of the task identities stored in cache the same as the cardinality of task_ident_key and wait_queue_key ? If I am correct there, why cannot we just use the result key instead of generating a uuid (which would mean fewer variables to move around)?

Reply:
Ok, thanks for the detailed explanation. It is almost entirely in line with my understanding. So to rephrase the need of having a unique task id, this is what could happen if we did not have it:
TASK1: runs the process on key1
TASK1: runs GET on key1
TASK1: sets up the queue and starts executing the query and waits the response from the DB
TASK2: starts the process on key1 (same key)
TASK2: runs the GET and learns that it has to wait
TASK1: gets the response builds the notify queue and returns
.... a huge amount of time passes, longer than the timeout ....
TASK3: starts the process on key1
TASK3: runs GET on key1
TASK3: sets up the queue and starts executing the query and waits the response from the DB
TASK3: gets the response builds the notify queue and returns
TASK1: finally runs BLPOP. 

So in this case, if the key of of the notify queue did not include the task id, TASK1 would be consuming from the notify queue built by TASK3, thus stealing results from the clients that were waiting on the results produced by TASK3.
Is this correct.

Reply:
This is pretty close, with the exception that the last step is performed by task 2 and not task 1 (as it has to wait), and the order in the second block of commands (the ones after the timeout) is not strictly important, since the `BLPOP` could actually start _before_ task 3 begins to execute (or even gets a cache miss to start its part of the process) and block until it the notify queue is created, as long as the value can be generated by task 3 before the `BLPOP` initiated by task 1 hits timeout.

I guess another way to phrase it is that if the duration that the `BLPOP` is active is ever extends beyond the duration of the task that triggered it (or that task deadline, whichever is smallest) + the cache key timeout (as this unlocks the potential for the cache value to be regenerated) we end up in dangerous territory since the `BLPOP` could attempt to remove a response that is not associated with it's "request".

I think that this is extremely unlikely to happen in reality but computers (and especially virtual machines) are weird and I'd rather play it safe unless there's a compelling reason not to. Though I guess the worst case here isn't catastrophic (some requests end up hitting timeout when they shouldn't) either so maybe it's not that big of a deal.

Reply:
This was about as clear of an explanation as I could think of about the importance of the task identity in the notify queue without going into more detail, let me know if this is useful: https://github.com/getsentry/snuba/blob/b689e2dca4366b1b3d86425fcd1d0409697ab0b3/snuba/state/cache/redis/backend.py#L93-L104